IaC config v1.3 documentation (#37097)
DataDog/documentation: content/en/security/code_security/iac_security/configuration.md

    text: "IaC Security Rules"
---

> Infrastructure as Code (IaC) Security detects IaC misconfigurations. By default, IaC Security scans repositories with [all supported rules][3]. You can customize which rules run and on which paths, as well as their severities, categories, and platforms. Configure these settings under the `iac` key in the Code Security configuration, either in Datadog or in a `code-security.datadog.yaml` file.

Reviewer: ```suggestion
Infrastructure as Code (IaC) Security detects IaC misconfigurations. By default, IaC Security scans repositories with [all supported rules][3]. You can customize which rules run and on which paths, as well as their severities and categories. Configure these settings under the `iac` key in the Code Security configuration, either in Datadog or in a `code-security.datadog.yaml` file.
```
I think that adding platforms here implies that you can change a rule's platform

    # Report only findings in these categories.
    only-categories:
      - "Encryption"
    # Do not run rules from these platforms.
    ignore-platforms:
      - Dockerfile
    # Only run rules from these platforms.
    only-platforms:
      - Terraform
      - Kubernetes
  # Per-rule configurations.
  rule-configs:
    terraform-aws-s3-bucket-without-encryption:
      ignore-paths:
        - "test/"
      severity: low
    kubernetes-deployment-without-resource-limits:
      only-paths:
        - "k8s/production/"
{{< /code-block >}}

The `iac` key supports the following fields:

| **Property** | **Type** | **Description** |
| --- | --- | --- |

> | `ignore-rules` | Array | A list of rule IDs that are ignored in scans. |

Reviewer 2: Original version is more concise. Defer to you on if there's a particular reason for the change made.

```suggestion
| `ignore-rules` | Array | A list of rule IDs to ignore. |
```

    # Report only findings in these categories.
    only-categories:
      - "Encryption"
    # Do not run rules from these platforms (v1.3+).
    ignore-platforms:
      - Dockerfile
    # Only run rules from these platforms (v1.3+).
    only-platforms:
      - Terraform
      - Kubernetes
  # Per-rule configurations (v1.3+).
  rule-configs:
    terraform-aws-s3-bucket-without-encryption:
      ignore-paths:
        - "test/"
      severity: low
    kubernetes-deployment-without-resource-limits:
      only-paths:
        - "k8s/production/"
{{< /code-block >}}

The `iac` key supports the following fields:

| **Property** | **Type** | **Description** |
| --- | --- | --- |

> | `ignore-rules` | Array | A list of rule IDs that will not be applied in scans. |

Reviewer: ```suggestion
| `ignore-rules` | Array | A list of rule IDs that are ignored in scans. |
```

> | `use-rules` | Array | A list of rule IDs that will be applied in scans. If specified, only these rules will be used (unless other configuration causes them to be ignored). |

Reviewer: ```suggestion
| `use-rules` | Array | A list of rule IDs that are applied in scans. If specified, _only_ these rules are applied in scans. `ignore-rules` takes precedence over `use-rules` (e.g. a rule that is in both arrays will be ignored).|
```

Reviewer: Not sure if this is true, but suggesting based on original wording

| --- | --- | --- |
| `only-paths` | Array | File paths or glob patterns. Only matching files are analyzed. |
| `ignore-paths` | Array | File paths or glob patterns to exclude. Matching files are not analyzed. |
| `only-severities` | Array | Severity levels to report. Findings with other severities are not reported. |
| `ignore-severities` | Array | Severity levels to ignore. |
| `only-categories` | Array | Categories to report. Findings in other categories are not reported. |
| `ignore-categories` | Array | Categories to ignore. |

> | `ignore-platforms` | Array | Platforms to skip. Rules from these platforms are not applied in scans. |

Reviewer 2: ```suggestion
| `ignore-platforms` | Array | Platforms to skip. Rules from these platforms are not applied. |
```

- `Supply-Chain`

{{< code-block lang="yaml" >}}
schema-version: v1.3
iac:
  global-config:
    ignore-categories:
      - "Access Control"
      - "Best Practices"
{{< /code-block >}}

### Platforms

Use `ignore-platforms` to skip specific platforms. Use `only-platforms` to restrict scanning to specific platforms.

**Possible values:**

- `Ansible`
- `CICD`
- `CloudFormation`
- `Dockerfile`
- `Kubernetes`
- `Terraform`

{{< code-block lang="yaml" >}}
schema-version: v1.3
iac:
  global-config:
    only-platforms:
      - Terraform
      - Kubernetes
{{< /code-block >}}

## Per-rule configuration

Use `rule-configs` to configure individual rules.

Each key under `rule-configs` is a rule ID. The following properties are supported per rule:

| **Property** | **Type** | **Description** |
| --- | --- | --- |
| `only-paths` | Array | File paths or glob patterns. The rule is applied only to files matching these patterns. |
| `ignore-paths` | Array | File paths or glob patterns to exclude. The rule is not applied to files matching these patterns. |

> | `severity` | String | Change the severity of findings generated by this rule. Accepted values: `critical`, `high`, `medium`, `low`, `info`. |

Reviewer 2: ```suggestion
| `severity` | String | Overrides the severity of findings generated by this rule. Accepted values: `critical`, `high`, `medium`, `low`, `info`. |
```

### Per-rule path scoping

Exclude a rule from certain paths, or restrict it to specific paths:

{{< code-block lang="yaml" >}}
schema-version: v1.3
iac:
  rule-configs:
    terraform-aws-s3-bucket-without-encryption:
      # Do not apply this rule in test directories.
      ignore-paths:
        - "test/"
        - "**/testdata/"
    kubernetes-deployment-without-resource-limits:
      # Apply this rule only in production manifests.
      only-paths:
        - "k8s/production/"
{{< /code-block >}}

Path patterns support glob syntax (`*`, `**`, `?`). Paths are relative to the repository root.

### Per-rule severity override

Change the severity of findings generated by a specific rule:

{{< code-block lang="yaml" >}}
schema-version: v1.3
iac:
  rule-configs:
    terraform-aws-s3-bucket-without-encryption:
      severity: low
{{< /code-block >}}

> The configured severity is used for all findings generated by that rule.

Reviewer 2: ```suggestion
This severity applies to all findings generated by that rule.
```
